src/osiris_log.erl

```
@@ -1884,6 +1821,72 @@ open_index_read(File) ->
+                    Result
+            end
+    end.
+
+scan_idx(Fd, Offset, #chunk_info{id = LastChunkInSegmentId, num = LastChunkInSegmentNum}) ->
+    case file:read(Fd, ?INDEX_RECORD_SIZE_B) of
+        {ok,
+         <<ChunkId:64/unsigned,
+           _Timestamp:64/signed,
+           _Epoch:64/unsigned,
+           FilePos:32/unsigned>>} ->
+            case Offset < ChunkId of
+                true ->
+                    %% offset is lower than the first chunk in this segment
+                    %% shouldn't really happen as we check the range above
+                    offset_out_of_range;
+                false ->
+                    case Offset > (LastChunkInSegmentId + LastChunkInSegmentNum - 1) of
+                        true ->
+                            %% offset higher than the last chunk in this segment
+                            {LastChunkInSegmentId + LastChunkInSegmentNum, eof};
+                        false ->
+                            %% offset is in this segment
+                            scan_idx(Fd, Offset, {ChunkId, FilePos})
+                    end
+            end;
+        eof ->
+            %% this should never happen - offset is in the range and we are raeding the first record
+            {error, eof};
+        {error, Posix} ->
+            Posix
+    end;
+
+scan_idx(Fd, Offset, PreviousChunk) ->
+    case file:read(Fd, ?INDEX_RECORD_SIZE_B) of
+        {ok,
+         <<ChunkId:64/unsigned,
+           _Timestamp:64/signed,
+           _Epoch:64/unsigned,
+           FilePos:32/unsigned>>} ->
+            case Offset < ChunkId of
+                true ->
+                    %% offset we are looking for is higher or equal to the start of the previous chunk
+                    %% but lower than the start of the current chunk -> return the previous chunk
+                    PreviousChunk;
+                false ->
+                    scan_idx(Fd, Offset, {ChunkId, FilePos})
+            end;
+        eof ->
```

should we return any errors to the caller here?

---

I don't think so - this is not an error but rather a normal situation where the offset we are looking for is in the last chunk but is not the first offset in that chunk.